base/libgit2/callbacks.jl

```diff
@@ -150,40 +183,46 @@ end
 
 function authenticate_userpass(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload)
     creds = Base.get(p.credential)::UserPasswordCredentials
+    modified = false
 
     # Reset password on sucessive calls
     if !p.first_pass
         creds.pass = ""
     end
 
-    if p.allow_prompt
-        username = creds.user
-        userpass = creds.pass
-        if isempty(username) || isempty(userpass)
-            prompt_url = git_url(scheme=p.scheme, host=p.host)
-            if Sys.iswindows()
-                response = Base.winprompt(
-                    "Please enter your credentials for '$prompt_url'", "Credentials required",
-                    isempty(username) ? p.username : username; prompt_username=true)
-                isnull(response) && return user_abort()
-                username, userpass = unsafe_get(response)
-            else
-                response = Base.prompt("Username for '$prompt_url'",
-                    default=isempty(username) ? p.username : username)
-                isnull(response) && return user_abort()
-                username = unsafe_get(response)
+    if p.first_pass && isfilled(creds)
+        modified = true
+    end
 
-                prompt_url = git_url(scheme=p.scheme, host=p.host, username=username)
-                response = Base.prompt("Password for '$prompt_url'", password=true)
-                isnull(response) && return user_abort()
-                userpass = unsafe_get(response)
-                isempty(userpass) && return user_abort()  # Ambiguous if EOF or newline
-            end
+    if p.remaining_prompts > 0 && (!modified || !isfilled(creds))
+        prompt_url = git_url(scheme=p.scheme, host=p.host)
+        if Sys.iswindows()
+            response = Base.winprompt(
+                "Please enter your credentials for '$prompt_url'", "Credentials required",
+                isempty(creds.user) ? p.username : creds.user; prompt_username=true)
```

Comment:
I'd be inclined to move the ternary out of the if-block and put the remaining code into an `user_auth_prompt` function.